helix-term/src/ui/editor.rs

```diff
@@ -375,12 +397,27 @@ impl EditorView {
                             }
                         } else {
                             let grapheme = Cow::from(grapheme);
+                            let is_whitespace;
 
                             let (grapheme, width) = if grapheme == "\t" {
+                                is_whitespace = true;
                                 // make sure we display tab as appropriate amount of spaces
                                 let visual_tab_width = tab_width - (visual_x as usize % tab_width);
-                                (&tab[..visual_tab_width], visual_tab_width)
+                                let grapheme_tab_width = tab
+                                    .grapheme_indices(true)
+                                    .nth(visual_tab_width)
+                                    .map(|(i, _)| i)
+                                    .unwrap_or(tab.len());
+
+                                (&tab[..grapheme_tab_width], visual_tab_width)
+                            } else if whitespace.render.space() == WhitespaceRenderValue::All
+                                && grapheme == " "
```

Comment:
This part of hte conditional should go first since most graphemes won't be space